src/Microsoft.ML.Tokenizers/Model/BertTokenizer.cs

```diff
@@ -0,0 +1,729 @@
+            }
+
+            if (tokenIds1 is not null)
+            {
+                foreach (int id in tokenIds1)
+                {
+                    buffer[written++] = id == ClsTokenId || id == SepTokenId || id == PadTokenId || id == MaskTokenId || id == UnknownTokenId ? 1 : 0;
+                }
+            }
+
+            return OperationStatus.Done;
+        }
+
+        /// <summary>
+        /// Create a mask from the two sequences passed to be used in a sequence-pair classification task. A BERT sequence pair mask has the following format:
+        ///         0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1
+        ///         | first sequence    | second sequence |
+        /// If <paramref name="tokenIds1"/> is null, this method only returns the first portion of the type ids (0s).
+        /// </summary>
+        /// <param name="tokenIds0">List of token IDs for the first sequence.</param>
+        /// <param name="tokenIds1">Optional list of token IDs for the second sequence.</param>
+        /// <returns>List of token type IDs according to the given sequence(s).</returns>
+        /// <exception cref="ArgumentNullException">When <paramref name="tokenIds0"/> is null.</exception>
+        public IReadOnlyList<int> CreateTokenTypeIdsFromSequences(IEnumerable<int> tokenIds0, IEnumerable<int>? tokenIds1 = null)
+        {
+            if (tokenIds0 is null)
+            {
+                throw new ArgumentNullException(nameof(tokenIds0));
+            }
+
+            // Add 2 for [CLS] and [SEP] tokens. Add 1 for [SEP] token if tokenIds1 is not null.
+            int capacity = tokenIds0.Count() + 2 + (tokenIds1 is null ? 0 : tokenIds1.Count() + 1);
+
+            List<int> typeIds = new List<int>(capacity);
+            for (int i = 0; i < tokenIds0.Count() + 2; i++) // Add 2 for [CLS] and [SEP] tokens.
+            {
+                typeIds.Add(0);
+            }
+
+            if (tokenIds1 is not null)
+            {
+                for (int i = 0; i < tokenIds1.Count() + 1; i++) // Add 1 for [SEP] token.
+                {
+                    typeIds.Add(1);
+                }
+            }
+
+            return typeIds;
+        }
+
+        public OperationStatus CreateTokenTypeIdsFromSequences(IEnumerable<int> tokenIds0, Span<int> buffer, out int written, IEnumerable<int>? tokenIds1 = null)
+        {
+            if (tokenIds0 is null)
+            {
+                throw new ArgumentNullException(nameof(tokenIds0));
+            }
+
+            written = 0;
+
+            // Add 2 for [CLS] and [SEP] tokens. Add 1 for [SEP] token if tokenIds1 is not null.
+            int capacity = tokenIds0.Count() + 2 + (tokenIds1 is null ? 0 : tokenIds1.Count() + 1);
+            if (buffer.Length < capacity)
+            {
+                return OperationStatus.DestinationTooSmall;
+            }
+
+            for (int i = 0; i < tokenIds0.Count() + 2; i++) // Add 2 for [CLS] and [SEP] tokens.
```

Comment:
This could be O(N^2) if tokenIds0 isn't ICollection.

All of the uses of Count() in this PR need to be revisted.

Reply:
I'll revisit all Count calls and fix them.